src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/GridGeometryProvider.java

```diff
@@ -0,0 +1,395 @@
+        }
+    }
+
+    /**
+     * Initialize the query to get granules, based on provided filter and region of interest (if any)
+     * 
+     * @param granules
+     * @return
+     * @throws TransformException
+     * @throws FactoryException
+     * @throws IOException
+     */
+    private Query initQuery(GranuleSource granules)
+            throws TransformException, FactoryException, IOException {
+        List<Filter> filters = new ArrayList<Filter>();
+
+        Query query = Query.ALL;
+
+        // Set bbox query if a ROI has been provided
+        if (roiManager != null) {
+            CoordinateReferenceSystem targetCRS = roiManager.getTargetCRS();
+            GeometryDescriptor geomDescriptor = granules.getSchema().getGeometryDescriptor();
+            CoordinateReferenceSystem indexCRS = geomDescriptor.getCoordinateReferenceSystem();
+            ReferencedEnvelope envelope = null;
+            if (targetCRS != null && !roiManager.isRoiCrsEqualsTargetCrs()) {
+                envelope = new ReferencedEnvelope(
+                    roiManager.getSafeRoiInTargetCRS().getEnvelopeInternal(), targetCRS);
+                MathTransform reprojectionTrasform = null;
+                if (!CRS.equalsIgnoreMetadata(targetCRS, indexCRS)) {
+                    reprojectionTrasform = CRS.findMathTransform(targetCRS, indexCRS, true);
+                    if (!reprojectionTrasform.isIdentity()) {
+                        envelope = envelope.transform(indexCRS, true);
+                    }
+                }
+            } else {
+                envelope = new ReferencedEnvelope(roiManager.getSafeRoiInNativeCRS().getEnvelopeInternal(), indexCRS);
+            }
+
+            final PropertyName geometryProperty = FeatureUtilities.DEFAULT_FILTER_FACTORY
+                    .property(geomDescriptor.getName());
+            filters.add(FeatureUtilities.DEFAULT_FILTER_FACTORY.bbox(geometryProperty, envelope));
+        }
+
+        // Add the filter if specified
+        if (filter != null) {
+            filters.add(filter);
+        }
+
+        // Set the query filter
+        if (!filters.isEmpty()) {
+            query = new Query();
+            query.setFilter(FeatureUtilities.DEFAULT_FILTER_FACTORY.and(filters));
+        }
+        return query;
+    }
+
+    /**
+     * Default GridGeometry retrieval based on native resolution.
+     * 
+     * @return
+     * @throws TransformException
+     * @throws IOException
+     */
+    private GridGeometry2D getNativeResolutionGridGeometry()
+            throws TransformException, IOException {
+        final ReferencedEnvelope roiEnvelope = roiManager != null
+                ? new ReferencedEnvelope(roiManager.getSafeRoiInNativeCRS().getEnvelopeInternal(),
+                        roiManager.getNativeCRS()) : null;
+        ScaleToTarget scaling = new ScaleToTarget(reader, roiEnvelope);
+        scaling.setTargetSize(null, null);
```

Comment:
Do we actually need to call this method? I have the impression it does not actually end up changing the state of the ScaleToTarget object.